app/src/org/commcare/xml/AndroidBulkCaseXmlParser.java

```diff
@@ -0,0 +1,120 @@
+import org.javarosa.xml.util.InvalidStructureException;
+import org.kxml2.io.KXmlParser;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+
+/**
+ * A bulk processing parser for the android platform. Provides superior performance when
+ * processing high case loads during syncing and/or processing.
+ *
+ * @author ctsims
+ */
+public class AndroidBulkCaseXmlParser extends BulkProcessingCaseXmlParser {
+    private HttpRequestEndpoints generator;
+    private final EntityStorageCache mEntityCache;
+    private final AndroidCaseIndexTable mCaseIndexTable;
+    private final SqlStorage<ACase> storage;
+
+    public AndroidBulkCaseXmlParser(KXmlParser parser,
+                                         SqlStorage<ACase> storage,
+                                         HttpRequestEndpoints generator) {
+        this(parser, storage, new EntityStorageCache("case"), new AndroidCaseIndexTable(), generator);
+    }
+
+    public AndroidBulkCaseXmlParser(KXmlParser parser,
+                                    SqlStorage<ACase> storage,
+                                    EntityStorageCache entityStorageCache,
+                                    AndroidCaseIndexTable indexTable,
+                                    HttpRequestEndpoints generator) {
+        super(parser);
+        this.generator = generator;
+        mEntityCache = entityStorageCache;
+        mCaseIndexTable = indexTable;
+        this.storage = storage;
+    }
+
+    protected SQLiteDatabase getDbHandle() {
+        return CommCareApplication.instance().getUserDbHandle();
+    }
+
+    @Override
+    protected Case buildCase(String name, String typeId) {
+        return new ACase(name, typeId);
+    }
+
+    @Override
+    protected void performBulkRead(Set<String> currentBulkReadSet, Map<String, Case> currentOperatingSet) throws InvalidStructureException, IOException, XmlPullParserException {
+        SQLiteDatabase db;
+        db = getDbHandle();
+        db.beginTransaction();
+        try {
+            for (ACase c : storage.getBulkRecordsForIndex(Case.INDEX_CASE_ID, currentBulkReadSet)) {
+                currentOperatingSet.put(c.getCaseId(), c);
+            }
+            db.setTransactionSuccessful();
+        } finally {
+            db.endTransaction();
+        }
+    }
+
+    @Override
+    protected void performBulkWrite(LinkedHashMap<String, Case> writeLog) throws IOException {
+        SQLiteDatabase db;
+        db = getDbHandle();
+        db.beginTransaction();
+        ArrayList<Integer> recordIdsToWipe = new ArrayList<>();
+        try {
+            for(String cid : writeLog.keySet()) {
+                Case c = writeLog.get(cid);
+                storage.write(c);
+                recordIdsToWipe.add(c.getID());
```

Comment:
They are getting "Wiped" from the case index and entity cache tables in order to get rebuilt. We don't bother "updating" index rows, we just delete all rows for a case and add new ones each time ("update if different" is just annoyingly complicated and not worth the performance gains since it involves a read)

In order to do that with maximal efficiency we do the wipe as one DB operation by putting them in a list.